### PR TITLE
Fixed Go, Ruby, and Python CLI Plugin output generation

### PR DESCRIPTION
**Pull request by leo**
### Related Issues

This contributes to resolving https://github.com/vercel/runtimes/issues/294 by aligning all CLI Plugins within this repo with how the File System API expects them to work, and resolves a number of bugs as part of that.

At the moment, API Routes overwrite each other's launcher files, so if you have more than one API Route per CLI Plugin, that would cause a failure when deployed. That's fixed now.

`vercel-plugin-node` wasn't yet updated, because it requires changes that are a little different and more complicated, because it requires multiple launcher files (a bridge and a launcher) per entrypoint. A PR for it will follow shortly.

Before this change, CLI Plugins would treat the entrypoint files (like `api/test.js`) as the Lambda's entrypoint file too, which is wrong, and it now has been changed to being the launcher file, with all the other files placed inside `inputs`.

CleanShot 2021-12-02 at 17 49 35@2x

#### Tests

- [x] The code changed/added as part of this PR has been covered with tests
- [x] All tests pass locally with `yarn test-unit`

#### Code Review

- [x] This PR has a concise title and thorough description useful to a reviewer
- [x] Issue from task tracker has a link to this PR

### CODE REVIEW

File: packages/build-utils/src/convert-runtime-to-plugin.ts [hunk 1]
```diff
@@ -90,8 +113,28 @@ export function convertRuntimeToPlugin(
...
+      // so we don't want to pollute this space unnecessarily. That means we have to clean
+      // up files that were created by the build, which is done further below.
+      const sourceFilesAfterBuild = await getSourceFiles(
+        workPath,
+        ignoreFilter
+      );
+
+      // Further down, we will need the filename of the Lambda handler
+      // for placing it inside `server/pages/api`, but because Legacy Runtimes
+      // don't expose the that filename directly, we have to construct it
```
Comment (AndyBitz) [inline packages/build-utils/src/convert-runtime-to-plugin.ts]:
```suggestion
      // don't expose the filename directly, we have to construct it
```

File: packages/build-utils/src/convert-runtime-to-plugin.ts [hunk 2]
```diff
@@ -141,12 +214,15 @@ export function convertRuntimeToPlugin(
...
+
       const json = JSON.stringify({
         version: 1,
-        files: tracedFiles.map(f => ({
-          input: normalizePath(relative(nft, f.absolutePath)),
-          output: normalizePath(f.relativePath),
-        })),
+        files: tracedFiles.map(file => {
+          return {
+            input: normalizePath(join('../../..', traceDirInputs, file)),
```
Comment (AndyBitz) [inline packages/build-utils/src/convert-runtime-to-plugin.ts]: I don't think this will work.

The reason why it was relative before, is because the file can be nested, e.g. `/api/users/info.py`, in that case it'd need to be `../../../../`, that's why `relative` was used.

File: packages/build-utils/test/unit.convert-runtime-to-plugin.test.ts [hunk 3]
```diff
@@ -68,12 +81,12 @@ describe('convert-runtime-to-plugin', () => {
...
           api: {
-            'index.py': expect.stringContaining('index'),
+            'index.py': expect.stringContaining('handler'),
             'index.py.nft.json': expect.stringContaining('{'),
             users: {
-              'get.py': expect.stringContaining('get'),
+              'get.py': expect.stringContaining('handler'),
               'get.py.nft.json': expect.stringContaining('{'),
-              'post.py': expect.stringContaining('post'),
+              'post.py': expect.stringContaining('handler'),
```
Comment (styfle) [inline packages/build-utils/test/unit.convert-runtime-to-plugin.test.ts:89]: Is this change necessary? Looks like a bug to me since its not reading the contents

File: packages/build-utils/src/convert-runtime-to-plugin.ts (hunk 2 above)
Comment (leo) [reply packages/build-utils/src/convert-runtime-to-plugin.ts]: Oh, right, I missed that.

Unfortunately `relative` adds one level too much, but I'll pick a different solution then!

File: packages/build-utils/src/convert-runtime-to-plugin.ts [hunk 4]
```diff
@@ -29,6 +29,24 @@ const shouldIgnorePath = (
...
 
+const getSourceFiles = async (workPath: string, ignoreFilter: any) => {
+  const list = await glob('**', {
+    cwd: workPath,
+  });
+
+  // We're not passing this as an `ignore` filter to the `glob` function above,
+  // so that we can re-use exactly the same `getIgnoreFilter` method that the
+  // Build Step uses (literally the same code). Note that this exclusion only applies
+  // when deploying. Locally, another exclusion further below is needed.
```
Comment (styfle) [inline packages/build-utils/src/convert-runtime-to-plugin.ts]: What does "futher below" mean? Seems like there is nothing below this loop

File: packages/build-utils/test/unit.convert-runtime-to-plugin.test.ts (hunk 3 above)
Comment (leo) [reply packages/build-utils/test/unit.convert-runtime-to-plugin.test.ts:89]: Yes, because the files placed in `server/pages/api` should be the actual Lambda entrypoints, and therefore the launcher files, not the request handlers that the user provided in their source code (those are separate files invoked by the launcher, depending on what Legacy Runtime is used).

File: packages/build-utils/src/convert-runtime-to-plugin.ts (hunk 4 above)
Comment (leo) [reply packages/build-utils/src/convert-runtime-to-plugin.ts]: Great point. I moved that comment, but didn't update it.

Updated it in https://github.com/vercel/vercel/pull/7117/commits/7fbf4668cc74168d33be2fe74b76ad1def2ed91b!

File: packages/build-utils/src/convert-runtime-to-plugin.ts (hunk 2 above)
Comment (leo) [reply packages/build-utils/src/convert-runtime-to-plugin.ts]: Seems like my other changes made `relative` work properly again! 🤔 

Fixed in https://github.com/vercel/vercel/pull/7117/commits/ca890767bc574df2a58bb48fd0dc6ecc5a096d5b and https://github.com/vercel/vercel/pull/7117/commits/7c4baeaafaf41609f47c97a09f5e9647fd8b89ee.

**Review by AndyBitz — APPROVED** [review]
Would probably be useful to have one e2e test for this that deploys the .output of a test of this. We'll definitely need to check if the handler logic will continue to work in prod

**Comment by leo** [comment]
Agreed! I can add some shortly.

**Comment by leo** [comment]
Everything except for unit tests on Windows seems to be working! Unit tests on Ubuntu and macOS are working.
